### ISSUE TITLE
se_mobilefacenet vs Mobilenet  compare problem

### ISSUE THREAD

**Issue by c7934597** [leondgarse/Keras_insightface#8]
你好，我看到checkpoint裡面有這兩個h5模型，經由上一篇你指點後，我使用cos相似度比較，分別在app/圖片辨識/webcam測試，發現se_mobilefacenet都有比較好的效果，而且差距有點大，mobilenet時常有辨識輸出錯誤，並且mobilenet是15MB，se_mobilefacenet只有5M，速度上也有差距，因此我想試著重現se_mobilefacenet訓練，可是我自己修改一直無法重現，有兩個問題請求您幫助，謝謝。

1.請問為什麼最後這幾層會不一樣，是有什麼原因修改嗎？另外我該如何修改，重現最後層數呢？
我把mobile_facenet.py的GDC下面那段改成這樣對嗎？

        M = Conv2D(emb_shape, 1, use_bias=False, activation=None)(M)
        M = BatchNormalization(axis=channel_axis)(M)
        M = PReLU(shared_axes=[1, 2])(M)
        if dropout > 0 and dropout < 1:
            M = Dropout(dropout)(M)
        M = Flatten()(M)
        M = Dense(emb_shape, use_bias=False, activation=None)(M)
        M = BatchNormalization(axis=channel_axis, name="embedding")(M)

se_mobilefacenet的最後幾層
![2020-09-19 21-20-18 的螢幕擷圖](https://user-images.githubusercontent.com/9107549/93668655-66961c80-fac0-11ea-87b5-e530a4898297.png)

mobilenet的最後幾層
![2020-09-19 21-21-25 的螢幕擷圖](https://user-images.githubusercontent.com/9107549/93668687-bb399780-fac0-11ea-9487-283ca2da9622.png)

2.

因為我想用其他資料集測試，我用以下訓練流程，可以重現你的訓練方式嗎? 或是哪邊需要修改變更呢?
```
from tensorflow import keras
from backbones import mobile_facenet
import losses
import train
import tensorflow_addons as tfa

basic_model = mobile_facenet.mobile_facenet(256, dropout=0, name="se_mobile_facenet_256", use_se=True)
data_path = '../datasets/faces_emore_112x112_folders'
eval_paths = ['../datasets/faces_emore/lfw.bin', '../datasets/faces_emore/cfp_fp.bin', '../datasets/faces_emore/agedb_30.bin']

tt = train.Train(data_path, save_path='keras_mobile_facenet_emore.h5', eval_paths=eval_paths, basic_model=basic_model, lr_base=0.001, batch_size=64, random_status=0, output_wd_multiply=10)
optimizer = tfa.optimizers.AdamW(weight_decay=5e-5)
sch = [
    {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "epoch": 20, "optimizer": optimizer},
    {"loss": losses.ArcfaceLoss(), "bottleneckOnly": True, "epoch": 2},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.35, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.3, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.25, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.2, "epoch": 20},
]

tt.train(sch, 0)
```

**Comment by leondgarse** [comment]
不好意思，最近有些其他事情在忙
1. 关于 `mobilenet` 与 `mobilefacenet` 最后几层的区别
- `mobilefacenet` 是参照其他项目中的实现修改的，所以保留了原来最后几层的实现
- `mobilenet` 最后几层其实跟 `mobilefacenet` 的差不多，只是将 `Dense` 替换成了 `Conv2D`，`Dense` 保留了 `bias`，参数量与计算差别不大
2. 上传的两个模型 `mobilefacenet` 与 `mobilenet`
-  `se_mobilefacenet` 是以前比较早训练的了，那时候还没有使用 `optimizer weight_decay`，还添加了 `dropout=0.4` 等，应该不是最优的
- `mobilenet` 是最近训练上传的，方式跟你的脚本差不多，`batch_size` 使用的 `512`
- 在识别效果上，我自己测试两个模型应该差别不大，但 `se_mobilefacenet` 虽然小，但速度慢很多，可以参照我这里的[测试结果](https://github.com/leondgarse/Keras_insightface#tflite-model-inference-time-test-on-arm32)
3. 看起来你的可能是 `batch_size` 太小了
4. 最近刚测试了一些更新，稍等下我上传一下，添加了一些参数以及与原版 `mxnet` 的对比，并且可以不再需要使用 `bottleneckOnly` 训练，最后几层参照原版的也有些调整

**Comment by c7934597** [comment]
不會，你的這些資源幫助我很多。
我有看到測試結果的表，mobilenet_v3這部分速度看起來很快，不過準確率會不會掉很多呢?
另外有看到xnn，我查看以下官方模型轉換方式，是哪個方式呢?

https://www.tensorflow.org/lite/performance/post_training_quantization

![123](https://user-images.githubusercontent.com/9107549/94402254-65d74780-019e-11eb-897c-64267f719e9a.png)

**Comment by leondgarse** [comment]
- `mobilenet_v3` 的我还没有训练过，只是测试了速度
- XNNPACK 使用的是 [float16量化方式](https://www.tensorflow.org/lite/performance/post_training_quantization#float16_quantization)，这里的速度测试是用的 [TFLite Model Benchmark Tool](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/lite/tools/benchmark)
- 似乎是因为 `tensorflow 2` 中 `tensor` 的修改，导致模型的一些操作变成了 `dynamic shape`，在使用 XNN 做前向的时候经常会报错，我通常是在 `tensorflow 1.15` 中转化 `float16`，此时还需要把 `h5` 模型转化成 `tf 1.15` 兼容的格式，以后可能会修复
- `tf 2` 中加载模型并保存为 `weights` + `json model`
```py
tf.__version__
# '2.3.0'

from tensorflow import keras
mm = keras.models.load_model('basic_model.h5')
mm.save_weights("weights_only.h5")
json_config = mm.to_json()
with open('model_config.json', 'w') as json_file:
    json_file.write(json_config)
```
- 修改 `model_config.json`, 删除不兼容的参数
```py
# For tf15 / tf20 saved json file, delete '"ragged": false,'
!sed -i 's/"ragged": false, //' model_config.json
# For tf-nightly saved json file, also replace '"class_name": "Functional"' by '"class_name": "Model"'
!sed -i 's/"class_name": "Functional"/"class_name": "Model"/' model_config.json
# For tf23 saved json file, also delete '"groups": 1, '
!sed -i 's/"groups": 1, //g' model_config.json
```
- `tensorflow 1.15` 中加载模型并保存 `h5`
```py
tf.__version__
# '1.15.0'

from tensorflow import keras
with open('model_config.json') as json_file:
    json_config = json_file.read()
new_model = keras.models.model_from_json(json_config)
new_model.load_weights('weights_only.h5')
new_model.save('basic_model_tf15.h5')
```
- `tensorflow 1.15` 中 `float16` 量化
```py
tf.__version__
# '1.15.0'
converter = tf.lite.TFLiteConverter.from_keras_model_file('basic_model_tf15.h5')
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_types = [tf.float16]

tflite_fp16_model = converter.convert()
open("model_f16.tflite", "wb").write(tflite_fp16_model)
```
